### PR TITLE
w249619149 Running One-off Commands

### PR DESCRIPTION
**Pull request by whatisdot**
Closes https://www.wrike.com/open.htm?id=249619149
Closes https://www.wrike.com/open.htm?id=252669067

This enables **both** one-off commands as well as an interactive console, using Docker and the AWS API.

### How it works
**Deployments**
Some adjustements were made to deployments. Now, we will be building a docker image locally, tagging it with the same string of text as the application version description, and pushing to ECR.
1. Generate the new version string to use.
1. Build docker image, tagging it with the new version.
1. Log in to ECR and push new image/tag.
1. Deploy to EB, noting the version being deployed.

**Running Command or Interactive Console**
Now that an image has been saved, you can run a command using it. This will:
1. Parse commands for EB environment name.
1. Find Id of the first EC2 instance running the current EB environment.
1. Find Id of the docker container running on the EC2 instance.
1. Get the environment variables of the container.
1. Log in to ECR, download, and run the command locally using the docker image.
 This will apply the environment variables that were previously pulled.
### Usage
**git_run_eb <environment_name> pwd**
```bash
[2018-07-09 10:52:06,568]: Using AWS CLI found at /usr/local/bin/aws
[2018-07-09 10:52:06,956]: Using default AWS profile found at ~/.aws/credentials
[2018-07-09 10:52:16,111]: Result:
/production
```

**git_run_eb <environment_name> ls -la**
```bash
[2018-07-09 10:52:24,183]: Using AWS CLI found at /usr/local/bin/aws
[2018-07-09 10:52:24,572]: Using default AWS profile found at ~/.aws/credentials
[2018-07-09 10:52:35,386]: Result:
total 64
drwxr-xr-x  9 root root 4096 Jul  5 16:12 .
drwxr-xr-x 22 root root 4096 Jul  9 15:52 ..
-rw-r--r--  1 root root 1468 Jul  4 19:24 Dockerfile
-rw-r--r--  1 root root 1351 Jul  4 19:24 Gemfile
-rw-r--r--  1 root root 9597 Jul  4 19:24 Gemfile.lock
-rw-r--r--  1 root root  227 Jul  4 19:24 Rakefile
drwxr-xr-x 10 root root 4096 Jul  4 19:24 app
drwxr-xr-x  2 root root 4096 Jul  4 19:24 bin
drwxr-xr-x  5 root root 4096 Jul  5 16:12 config
-rw-r--r--  1 root root  130 Jul  4 19:24 config.ru
drwxr-xr-x  3 root root 4096 Jul  4 19:24 db
drwxr-xr-x  4 root root 4096 Jul  4 19:24 lib
drwxr-xr-x  2 root root 4096 Jul  4 19:24 public
drwxr-xr-x  3 root root 4096 Jul  4 19:24 vendor
```

**git_run_eb <environment_name> rails c**
```sh
$ git_run_eb stage rails c
[2018-07-12 14:41:25,391]: Using AWS CLI found at /usr/local/bin/aws

[2018-07-12 14:41:25,781]: Using default AWS profile found at ~/.aws/credentials
/usr/local/bundle/gems/activesupport-5.0.1/lib/active_support/xml_mini.rb:51: warning: constant ::Fixnum is deprecated
/usr/local/bundle/gems/activesupport-5.0.1/lib/active_support/xml_mini.rb:52: warning: constant ::Bignum is deprecated
W, [2018-07-12T19:41:36.771996 #13]  WARN -- : You are setting a key that conflicts with a built-in method OmniAuth::AuthHash::InfoHash#name defined at /usr/local/bundle/gems/omniauth-1.3.2/lib/omniauth/auth_hash.rb:34. This can cause unexpected behavior when accessing the key via as a property. You can still access the key via the #[] method.
/usr/local/bundle/gems/activesupport-5.0.1/lib/active_support/core_ext/numeric/conversions.rb:138: warning: constant ::Fixnum is deprecated
/usr/local/bundle/gems/activejob-5.0.1/lib/active_job/arguments.rb:38: warning: constant ::Fixnum is deprecated
/usr/local/bundle/gems/activejob-5.0.1/lib/active_job/arguments.rb:38: warning: constant ::Bignum is deprecated
Running via Spring preloader in process 22
Loading stage environment (Rails 5.0.1)
irb(main):001:0>
```

### CODE REVIEW

File: lib/run_eb.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+  def get_instance_id(self, environment):
+    arguments = [
+      'elasticbeanstalk describe-environment-resources',
+      '--region us-west-2',
+      '--profile assumed_role_config',
+      '--environment-name ' + environment
+    ]
+    response = self.aws(arguments)
+    return response['EnvironmentResources']['Instances'][0]['Id']
```
Comment (jeet4320) [inline lib/run_eb.py]: what if someone wants to run one-off command on other instances

Comment (whatisdot) [reply lib/run_eb.py]: Why is that? It is presumed that instances are identical clones. If they aren't identical, it is an integrity problem. This is a separate concern.

**Comment by whatisdot** [comment]
Currently investigating alternate ways to execute remote commands. Putting Not Ready on this in the meantime.

**Comment by whatisdot** [comment]
@jeet4320 @rstaudt2 Here's my take on the interactive console and one-off commands solution. It requires Docker to be installed locally, but doesn't require the user to issue any Docker commands. Let me know what you think.

**Comment by whatisdot** [comment]
@jeet4320 @rstaudt2 Requesting LGs. Based on our conversation, we can merge this and follow up on ways to improve this process.

**Pull request merged**